test/common/router/config_impl_test.cc

```diff
@@ -1404,6 +1410,35 @@ class RouterMatcherCookieHashPolicyTest : public RouterMatcherHashPolicyTest {
   }
 };
 
+TEST_F(RouterMatcherCookieHashPolicyTest, EmptyKeyMatchesWholeCookieLine) {
+  // With no key specified, the whole cookie line is used.
+  firstRouteHashPolicy()->mutable_cookie()->set_name("");
+  {
+    // Nothing to hash on.
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    Router::RouteConstSharedPtr route = config().route(headers, 0);
+    EXPECT_FALSE(
+        route->routeEntry()->hashPolicy()->generateHash(nullptr, headers, add_cookie_nop_));
+  }
+  uint64_t hash_1;
+  uint64_t hash_2;
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    headers.addCopy("Cookie", "choco=late; su=gar");
+    Router::RouteConstSharedPtr route = config().route(headers, 0);
+    hash_1 =
+        route->routeEntry()->hashPolicy()->generateHash(nullptr, headers, add_cookie_nop_).value();
+  }
+  {
+    Http::TestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    headers.addCopy("Cookie", "bluh=blur");
```

Comment:
You could write an implementation of cookie hashing that used only the first cookie in the list (or only the last) and it would pass this test, which means your test isn't checking what you intend it to check.

Reply:
though this is reverted, but this test is saying:

* The key name is set to empty.

* if cookies header is empty:
    there is no hash.
else:
    there is hash and for different cookies headers, hashes are different. 

Do you have a better idea on how to express this?

Reply:
You would definitely want these test cases, but you should probably have an additional one that changes only one of the cookies from the second test and asserts that the hashes are different. That way, if you accidentally (or someone else maliciously) changed the implementation of whole-cookie header hashing to use only the first or the last cookie, your tests would no longer pass. As it is, though, there are a couple alternate and plausible implementations that would pass these tests.